### ISSUE TITLE
File with lower_case_with_underscores identifier message is persisting.

### ISSUE THREAD

**Issue by FMorschel** [Dart-Code/Dart-Code#5343]
After https://github.com/Dart-Code/Dart-Code/issues/5270 (I think this might be the cause), if you make a code like this (no actual file with this name):

```dart
export 'MyFile.dart';
```

You'll see the new message for it. But even if you replace this export completely (no second file needed) the message will persist.

> The file name 'MyFile.dart' isn't a lower_case_with_underscores identifier.
Try changing the name to follow the lower_case_with_underscores style.
You may need to close the file and restart VS Code after renaming a file by only casing.

This is persisted because the message is showing on a non-existing file.

![image](https://github.com/user-attachments/assets/9d49c07d-b8a8-4076-8b9b-d68a8efd81ca)

![image](https://github.com/user-attachments/assets/3d998dcd-c866-465d-a27a-f35351fabf63)

**Comment by DanTup** [comment]
I don't seem to be able to reproduce this without `MyFile.dart` existing:

![image](https://github.com/user-attachments/assets/d60971df-18e2-4f6a-a295-2ff9efa8e8fb)

Are you able to reproduce this in a clean project? Did `MyFile.dart` previously exist? (is it possible you deleted it from disk, but VS Code still has it "open but deleted" so we have an in-memory overlay for it?)

**Comment by FMorschel** [comment]
> Are you able to reproduce this in a clean project?

That was a clean project. 

> Did MyFile.dart previously exist?

Not to my knowledge. I could repro this with some different file names.

Are there any logs that would help with this?

**Comment by DanTup** [comment]
> Are there any logs that would help with this?

Yeah, the instrumentation log will help. But exact steps to reproduce in a new project would also help - it'll be easier to debug if I can repro it versus trying to reverse engineer the logs :-)

**Comment by FMorschel** [comment]
Forgot about posting these:

<details>

<summary><strong>Workspace Environment</strong></summary>

```text
Dart Code extension: 3.101.20241118
Flutter extension: 3.101.20241031 (activated)

App: Visual Studio Code
App Host: desktop
Version: win 1.95.3

Workspace type: Dart (LSP)
Workspace name: message_error

Dart (3.7.0-edge.d75c493e27f9bf2a89043860132462be665065c2): C:\Users\felip_0vh5fa6\dart
Flutter (3.24.4): C:\Users\felip_0vh5fa6\.puro\envs\stable\flutter (No device)

```

</details>

<details>

<summary><strong>Output from 'dart info'</strong></summary>

`C:\Users\felip_0vh5fa6\dart\bin\dart.exe info`

If providing this information as part of reporting a bug, please review the information
below to ensure it only contains things you're comfortable posting publicly.

#### General info

- Dart 3.7.0-edge.d75c493e27f9bf2a89043860132462be665065c2 (main) (Tue Nov 12 10:33:29 2024 +0000) on "windows_x64"
- on windows / "Windows 11 Pro" 10.0 (Build 22631)
- locale is pt-BR

#### Project info

- sdk constraint: '^3.7.0-edge.d75c493e27f9bf2a89043860132462be665065c2'
- dependencies: 
- dev_dependencies: lints, test

#### Process info

| Memory | CPU | Elapsed time | Command line |
| -----: | --: | -----------: | ------------ |
|   0 MB |  -- |              | dart.exe     |
|   0 MB |  -- |              | dart.exe     |
|   0 MB |  -- |              | dart.exe     |
|   0 MB |  -- |              | dart.exe     |
|   0 MB |  -- |              | dart.exe     |
|   0 MB |  -- |              | dart.exe     |
|   0 MB |  -- |              | dart.exe     |
|   0 MB |  -- |              | dart.exe     |
|   0 MB |  -- |              | dart.exe     |
|   0 MB |  -- |              | dart.exe     |
|   0 MB |  -- |              | dart.exe     |

</details>

New empty dart project:

https://github.com/user-attachments/assets/3f9ad36d-5ca6-4e4c-8f64-cb6b4c713747

Here are the full logs:

[message_error.zip](https://github.com/user-attachments/files/17815348/message_error.zip)

**Comment by DanTup** [comment]
Thanks! I can repro following your video.

Btw, your analysis seems to update incredibly slowly for such a small project (it takes a long time for the error about `e` being undefined to be updated). Is this the kind of performance you normally see? For me, it updates as fast as I type:

https://github.com/user-attachments/assets/356c8e22-84d7-49d8-b5b9-92e9d05116f9

**Comment by bwilkerson** [comment]
> Should we be computing any diagnostics and/or lints for files that don't exist?

I can't see any reason to do so.

> Should we be sending diagnostics for any files that don't exist? (I feel like we shouldn't, but I also worry whether we might leave stale diagnostics if we just skip this.. so we'd probably want to send an empty set if the file doesn't exist instead, but it feels clunky)

If we can detect that a file has been deleted (and I think we can) then we ought to be able to send an empty list of diagnostics at that point rather than every time we re-discover that the file no longer exists.

> Should `getResolvedUnit` even return a `ResolvedUnitResult` for non-existent files?

If it doesn't then we'd have to fix `DartFileEditBuilder` to work when there's no `ResolvedUnitResult`. Or find some other way to allow us to build the edit used to create the file with some given content.

@scheglov

**Comment by DanTup** [comment]
@bwilkerson 

> > Should getResolvedUnit even return a ResolvedUnitResult for non-existent files?
>
> If it doesn't then we'd have to fix DartFileEditBuilder to work when there's no ResolvedUnitResult. Or find some other way to allow us to build the edit used to create the file with some given content

Good point, probably we shouldn't change that then.

I can think of a few possible fixes:

1. Change the file_names lint to never produce a diagnostic for non-existent files (I suspect it's probably the only one that does, but it's not guaranteed now and forever)
2. Skip running lints for non-existent files (probably in `LibraryAnalyzer._computeLints()`?)
3. Change the analysis server to just send an empty set whenever it's sending diagnostics for a non-existent file (probably in `NotificationManager.sendAnalysisErrors()`?)

I felt like the last one would be best, because we don't have to worry about skipping sending diagnostics that might result in stale data (eg. if a file is deleted and then we never send an event to clear them on the client). However, `NotificationManager.sendAnalysisErrors` only works with paths so doesn't have the `ResolvedUnitResult` at this point (and in fact, there might not be one, since we can produce diagnostics for non-Dart files).

Any thoughts/opinions on the best way to fix?

@scheglov 

> The error that you imported a not existing file is not in the not existing file, so we don't need its `ResolvedUnitResult`. 

That's true, but we produce a fix ("Create file 'foo.dart'") which has some default content in it, so we use the `DartFileEditBuilder` with the `ResolvedUnitResult` for the non-existent file (we also produce the file_names diagnostic against it.. there are two diagnostics produced here - the one that you're importing a missing file, and one that the missing file that we analyzed has a name that violates the lint).

**Comment by scheglov** [comment]
For the analyzer it is not very useful to report diagnostics for a not existing file.
And this will also fix the observed issue with sending these diagnostics by DAS.

We don't have to use `DartFileEditBuilder`, AFAIK nothing in the protocol requires this, all we need is a `SourceFileEdit`. It looks that we have `addGenericFileEdit` that might be useful.

**Comment by bwilkerson** [comment]
> 1. Change the file_names lint ...

That does seem a little too special-case.

> 2. Skip running lints for non-existent files ...

That's nice and general, and might improve performance by skipping some work, though it presumably doesn't happen often, so meh.

> 3. Change the analysis server to just send an empty set whenever it's sending diagnostics for a non-existent file ...

I think that one would also be fine.

So I'm OK with either of the latter two.

I'm assuming that for both of these latter two cases we would ask the `FileState` for existence information because asking the resource manager would be too inefficient.

**Comment by DanTup** [comment]
> We don't have to use `DartFileEditBuilder`, AFAIK nothing in the protocol requires this, all we need is a `SourceFileEdit`. It looks that we have `addGenericFileEdit` that might be useful.

That's true, though changing `getResolvedUnit` to not return a `ResolvedUnitResult` for a non-existing file is probably a breaking change? (if we did this, we should probably deprecate/remove `exists` and update any code using it to expect `null` instead?).

I'll try implementing (2) for now, since I agree it's better to not produce them (or run the code) than just filter them out lately, and I guess we must already be handling flushing errors for deleted files reliably (and not relying on something triggering analysis of those now-deleted files).

> I'm assuming that for both of these latter two cases we would ask the `FileState` for existence information because asking the resource manager would be too inefficient.

Yep, I think so (because we also need it to be consistent and I'm not sure if resource manager would be guaranteed to be).

**Comment by DanTup** [comment]
I've opened https://dart-review.googlesource.com/c/sdk/+/398881.

**Comment by scheglov** [comment]
`CreateField` does not have to use `addDartFileEdit`, so I don't understand the comment about `getResolvedUnit`. We don't have to invoke it, right?

**Comment by DanTup** [comment]
> `CreateFile` does not have to use `addDartFileEdit`

Yeah, we could definitely change this fix to use a generic file edit builder and/or handle `getResolvedUnit()` returning `null` for non-existent files. We could also update all places that call `getResolvedUnit` and check `exists` to instead expect `null`. But if there is external code calling `getResolvedUnit()` and checking `exists` (or at least, handling non-null with `exists=false` differently to `null`), I think we should break them?

If you think we should do this (effectively making `exists` on `ResolvedUnitResult` redundant because it would always be `true` if I understand correctly?), I could try making the change to see if anything fails (and then we can decide how to release it if it's considered breaking)?

**Comment by scheglov** [comment]
Which `getResolvedUnit` are we discussing here? The one from `AnalysisSession`? Or from `AnalysisServer`?

The `AnalysisSession` one already returns `SomeResolvedUnitResult` that can be many results, not necessary `ResolvedUnitResult`.

Analyzing a not existing file is not very useful, so we could make it `NotExistingPathResult` if we want. Not sure how useful it is too though.

**Comment by DanTup** [comment]
Fixed by https://github.com/dart-lang/sdk/commit/bb4850bb8d42098bc29dea42a8a99f0051c2cfba - ships in a future SDK release.

**Comment by DanTup** [comment]
> Which `getResolvedUnit` are we discussing here? The one from `AnalysisSession`? Or from `AnalysisServer`?

Oh, I was looking at `AnalysisSession` but I see my comment about `null` was slightly incorrect because I was looking at code that did `session?.getResolvedUnit(path)` (so the `null` only came from the lack of a session, not the return value).

There aren't many places that check `exists` on `ResolvedUnitResult`:

![image](https://github.com/user-attachments/assets/27dc2bff-ffa5-49d7-a53f-49f0d61663a2)

So we could perhaps use a `NotExistingPathResult` and update those places. However...

I thought of another place currently using `addDartFileEdit` for potentially-missing files, and that's the "Move Top Level to File" refactor. You can choose either an existing file or a new one. Part of what it does is add `import`s, so I don't think we could use a generic file builder here.

I'm not sure what scale of changes it would need if a non-existent file no longer had a `ResolvedUnitResult` though - there do seem to be quite a few references to it in the edit builder (accessing things like `path`, `file`, `content`, `libraryElement`, `lineInfo`). I suspect most of them are trivial changes to handle it missing, but whether it's worth adding all those conditions in order to change this is less clear - I'm happy to try it out if you think it would be better.

**Comment by DanTup** [comment]
Yeah, this was the only issue I know of where analyzing missing files was causing a problem (because the lint is based only on the name and doesn't require any actual file/content to produce a diagnostic).

(This issue is fixed by bb4850bb8d42098bc29dea42a8a99f0051c2cfba, so there are now no _known_ issues - to me - as a result of this)
